internal/services/tesla_fleet_api_service.go

```diff
@@ -0,0 +1,143 @@
+	"github.com/rs/zerolog"
+	"golang.org/x/oauth2"
+)
+
+//go:generate mockgen -source tesla_fleet_api_service.go -destination mocks/tesla_fleet_api_service_mock.go
+type TeslaFleetAPIService interface {
+	CompleteTeslaAuthCodeExchange(ctx context.Context, authCode, redirectURI, region string) (*TeslaAuthCodeResponse, error)
+	GetVehicles(ctx context.Context, token, region string) ([]TeslaVehicle, error)
+}
+
+var teslaScopes = []string{"openid offline_access", "user_data", "vehicle_device_data", "vehicle_cmds", "vehicle_charging_cmds", "energy_device_data", "energy_device_data", "energy_cmds"}
+
+type GetVehiclesResponse struct {
+	Response []TeslaVehicle `json:"response"`
+}
+
+type TeslaFleetAPIError struct {
+	Error            string `json:"error"`
+	ErrorDescription string `json:"error_description"`
+	ReferenceID      string `json:"referenceId"`
+}
+
+type TeslaAuthCodeResponse struct {
+	AccessToken  string `json:"access_token"`
+	RefreshToken string `json:"refresh_token"`
+	IDToken      string `json:"id_token"`
+	Expiry       int    `json:"expiry"`
+	TokenType    string `json:"token_type"`
+}
+
+type teslaFleetAPIService struct {
+	Settings   *config.Settings
+	HTTPClient *http.Client
+	log        *zerolog.Logger
+}
+
+func NewTeslaFleetAPIService(settings *config.Settings, logger *zerolog.Logger) TeslaFleetAPIService {
+	return &teslaFleetAPIService{
+		Settings:   settings,
+		HTTPClient: &http.Client{},
+		log:        logger,
+	}
+}
+
+// CompleteTeslaAuthCodeExchange godoc
+// @Description Call Tesla Fleet API and exchange auth code for a new auth and refresh token
+// @Param       authCode - authorization code to exchange
+// @Param       redirectURI - redirect uri to pass on as part of the request to for oauth exchange
+// @Param       region - API region which is used to determine which fleet api to call
+// @Returns     {object} services.TeslaAuthCodeResponse
+func (t *teslaFleetAPIService) CompleteTeslaAuthCodeExchange(ctx context.Context, authCode, redirectURI, region string) (*TeslaAuthCodeResponse, error) {
+	conf := oauth2.Config{
+		ClientID:     t.Settings.Tesla.ClientID,
+		ClientSecret: t.Settings.Tesla.ClientSecret,
+		Endpoint: oauth2.Endpoint{
+			TokenURL: t.Settings.Tesla.TokenAuthURL,
+		},
+		RedirectURL: redirectURI,
+		Scopes:      teslaScopes,
+	}
+
+	ctxTimeout, cancel := context.WithTimeout(ctx, time.Second*30)
+	defer cancel()
+
+	tok, err := conf.Exchange(ctxTimeout, authCode, oauth2.SetAuthURLParam("audience", fmt.Sprintf("https://fleet-api.prd.%s.vn.cloud.tesla.com", region)), oauth2.SetAuthURLParam("grant_type", "authorization_code"))
+	if err != nil {
+		var e *oauth2.RetrieveError
+		errString := ""
+		if errors.As(err, &e) {
+			errString = e.ErrorDescription
+		}
+		return nil, fmt.Errorf("error occurred completing authorization: %s", errString)
+	}
+
+	return &TeslaAuthCodeResponse{
+		AccessToken:  tok.AccessToken,
+		RefreshToken: tok.RefreshToken,
+		Expiry:       int(tok.Expiry.Unix()),
```

Comment:
I guess we'll see how this gets used. I'm skeptical that the unix timestamp is going to be useful over the `time.Time.`